ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+					format!("Block is too far in the future, timestamp: {}, limit: {}", header.timestamp(), limit)
+				).into());
+			}
+		}
+
+		let is_checkpoint = header.number() % self.epoch_length == 0;
+
+		if is_checkpoint && *header.author() != NULL_AUTHOR {
+			return Err(Box::new("Checkpoint block must enforce zero beneficiary").into());
+		}
+
+		// Nonce must be 0x00..0 or 0xff..f
+		let nonce = header.decode_seal::<Vec<&[u8]>>()?[1];
+		if *nonce != NONCE_DROP_VOTE && *nonce != NONCE_AUTH_VOTE {
+			return Err(Box::new("nonce must be 0x00..0 or 0xff..f").into());
+		}
+		if is_checkpoint && *nonce != NULL_NONCE[..] {
+			return Err(Box::new("Checkpoint block must have zero nonce").into());
+		}
+
+		// Ensure that the extra-data contains a signer list on checkpoint, but none otherwise.
+		if (!is_checkpoint && header.extra_data().len() != (SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH))
+			|| (is_checkpoint && header.extra_data().len() <= (SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH))
+			|| (is_checkpoint && (header.extra_data().len() - (SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH)) % 20 != 0) {
+			return Err(Box::new(format!("Invalid extra_data length, got {}", header.extra_data().len())).into());
+		}
+
+		// Ensure that the mix digest is zero as we don't have fork protection currently
+		let mixhash = header.decode_seal::<Vec<&[u8]>>()?[0];
+		if mixhash != NULL_MIXHASH {
+			return Err(Box::new("mixhash must be 0x00..0 or 0xff..f.").into());
+		}
+
+		// Ensure that the block doesn't contain any uncles which are meaningless in PoA
+		if *header.uncles_hash() != NULL_UNCLES_HASH {
+			return Err(Box::new(format!(
+				"Invalid uncle hash, got: {}, expected: {}.",
+				header.uncles_hash(),
+				NULL_UNCLES_HASH,
+			)).into());
+		}
+
+		// Ensure that the block's difficulty is meaningful (may not be correct at this point)
+		if *header.difficulty() != DIFF_INTURN && *header.difficulty() != DIFF_NOTURN {
+			return Err(Box::new(format!(
+				"invalid difficulty: expected {} or {}, got: {}.",
+				DIFF_INTURN,
+				DIFF_NOTURN,
+				header.difficulty(),
+			)).into());
+		}
+
+		// All basic checks passed, continue to next phase
+		Ok(())
+	}
+
+	fn verify_block_unordered(&self, _header: &Header) -> Result<(), Error> {
+		// Nothing to check here.
+		Ok(())
+	}
+
+	/// Verify block family by looking up parent state (backfill if needed), then try to apply current header.
+	/// see https://github.com/ethereum/go-ethereum/blob/master/consensus/clique/clique.go#L338
+	fn verify_block_family(&self, header: &Header, parent: &Header) -> Result<(), Error> {
+		// Ignore genesis block.
+		if header.number() == 0 {
+			return Ok(());
+		}
+
+		// parent sanity check
+		if parent.hash() != *header.parent_hash() || header.number() != parent.number() + 1 {
+			return Err(Box::new("invalid parent").into());
+		}
+
+		// Ensure that the block's timestamp isn't too close to it's parent
+		if parent.timestamp() + self.period > header.timestamp() {
```

Comment:
use [checked_add](https://doc.rust-lang.org/std/primitive.u64.html#method.saturating_add) or [saturating_add](https://doc.rust-lang.org/std/primitive.u64.html#method.saturating_add) to prevent integer overflow in `parent.timestamp() + self.period`